src/sctools/reader.py

```diff
@@ -0,0 +1,122 @@
+import os
+import gzip
+import bz2
+from copy import copy
+from collections.abc import Iterable
+
+
+class Reader:
+    """
+    Basic reader object that seamlessly loops over multiple input files
+
+    Can be subclassed to create readers for specific file types (fastq, gtf, etc.)
+    """
+
+    def __init__(self, files='-', mode='r', header_comment_char=None):
+        """
+        :param list|str files: file or list of files to be read. Defaults to sys.stdin
+        :param mode: (Default 'r') returns string objects. Change to 'rb' to
+          return bytes objects.
+        """
+
+        if isinstance(files, str):
+            self._files = [files]
+        elif isinstance(files, Iterable):  # test items of iterable
+            files = list(files)
+            if all(isinstance(f, str) for f in files):
+                self._files = files
+            else:
+                raise TypeError('all passed files must be type str')
+        else:
+            raise TypeError('files must be a string filename or a list of such names.')
+
+        # set open mode:
+        if mode not in {'r', 'rb'}:
+            raise ValueError('mode must be one of r, rb')
+        self._mode = mode
+
+        if isinstance(header_comment_char, str) and mode == 'rb':
+            self._header_comment_char = header_comment_char.encode()
+        else:
+            self._header_comment_char = header_comment_char
+
+    @property
+    def filenames(self):
+        return self._files
+
+    def __len__(self):
+        """
+        return the length of the Reader object. this should typically not be used with sys.stdin,
+        as it will consume the input
+        """
+        return sum(1 for _ in self)
+
+    def __iter__(self):
+        for file_ in self._files:
+
+            # set correct mode
+            if self._mode == 'r' and any(file_.endswith(s) for s in ['.gz', '.bz2']):
```

Comment:
Why do we treat this as a special case? We already set mode = 'r' when self._mode == 'r' by default in the else. In this special case we set it to 'rt' instead. And according to this, r and rt are the same thing:
https://stackoverflow.com/questions/23051062/open-files-in-rt-and-wt-modes

Reply:
the `gzip` and `bzip2` libraries _do_ make a distinction between 'r' and 'rt'. It's very annoying: 

https://docs.python.org/3/library/gzip.html#gzip.open

Reply:
... if they have proper extensions. I will look into a better way of detecting compressing type.

Reply:
Okay, in that case I think what you have is perfectly reasonable here. Maybe just add a comment explaining that rt really does differ from r in this case.